packages/remote-feature-flag-controller/src/remote-feature-flag-controller.ts

```diff
@@ -0,0 +1,211 @@
+ */
+export class RemoteFeatureFlagController extends BaseController<
+  typeof controllerName,
+  RemoteFeatureFlagControllerState,
+  RemoteFeatureFlagControllerMessenger
+> {
+  readonly #fetchInterval: number;
+
+  #disabled: boolean;
+
+  #clientConfigApiService: AbstractClientConfigApiService;
+
+  #inProgressFlagUpdate?: Promise<FeatureFlags>;
+
+  /**
+   * Constructs a new RemoteFeatureFlagController instance.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The controller messenger used for communication.
+   * @param options.state - The initial state of the controller.
+   * @param options.clientConfigApiService - The service instance to fetch remote feature flags.
+   * @param options.fetchInterval - The interval in milliseconds before cached flags expire. Defaults to 1 day.
+   * @param options.disabled - Determines if the controller should be disabled initially. Defaults to false.
+   */
+  constructor({
+    messenger,
+    state,
+    clientConfigApiService,
+    fetchInterval = DEFAULT_CACHE_DURATION,
+    disabled = false,
+  }: {
+    messenger: RemoteFeatureFlagControllerMessenger;
+    state?: Partial<RemoteFeatureFlagControllerState>;
+    clientConfigApiService: AbstractClientConfigApiService;
+    fetchInterval?: number;
+    disabled?: boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: remoteFeatureFlagControllerMetadata,
+      messenger,
+      state: {
+        ...getDefaultRemoteFeatureFlagControllerState(),
+        ...state,
+      },
+    });
+
+    this.#fetchInterval = fetchInterval;
+    this.#disabled = disabled;
+    this.#clientConfigApiService = clientConfigApiService;
+  }
+
+  /**
+   * Checks if the cached feature flags are expired based on the fetch interval.
+   *
+   * @returns Whether the cache is expired (`true`) or still valid (`false`).
+   * @private
+   */
+  #isCacheExpired(): boolean {
+    return Date.now() - this.state.cacheTimestamp < this.#fetchInterval;
+  }
+
+  /**
+   * Retrieves the remote feature flags, fetching from the API if necessary.
+   * Uses caching to prevent redundant API calls and handles concurrent fetches.
+   *
+   * @returns A promise that resolves to the current set of feature flags.
+   */
+  async getRemoteFeatureFlags(): Promise<FeatureFlags> {
+    if (this.#disabled) {
+      return [];
+    }
+
+    if (this.#isCacheExpired()) {
+      return this.state.remoteFeatureFlags;
+    }
+
+    if (this.#inProgressFlagUpdate) {
+      return await this.#inProgressFlagUpdate;
+    }
+
+    const { promise, resolve } = createDeferredPromise<FeatureFlags>({
+      suppressUnhandledRejection: true,
+    });
+    this.#inProgressFlagUpdate = promise;
+
+    try {
+      const flags =
+        await this.#clientConfigApiService.fetchRemoteFeatureFlags();
+      if (flags.cachedData.length > 0) {
+        this.updateCache(flags.cachedData);
+        resolve(flags.cachedData);
+      }
+      return await promise;
```

Comment:
If there is no cached data, and if there is an API error, it looks like this promise will stay unresolved forever

Reply:
Should we have a `catch` here?

Reply:
1) Agreed ! when there is no cached data, we should return `[]` 
2) The error catch is handled in service side, hence we don't need catch here, hope it makes sense !
```
return {
        error: true,
        message: err.message || 'Unknown error',
        statusCode: err.response?.status?.toString() || null,
        statusText: err.response?.statusText || null,
        cachedData: cachedData || [],
        cacheTimestamp: cacheTimestamp || Date.now(),
      };
```
Addressed https://github.com/MetaMask/core/commit/276357dc6479e1c542d2c47d6a288a2177832257

Reply:
What was the reason for handling the error as is done in the service?

Reply:
I notice that the example gas price service does not handle errors this way (https://github.com/MetaMask/core/blob/main/examples/example-controllers/src/gas-prices-service/gas-prices-service.ts#L62)